File: Client/Source/LoginSystem_Client/Private/Utils/Cryptography.cpp
#include "Utils/Cryptography.h"
#include "Misc/Base64.h"

#include <iomanip>
#include <sstream>
#include <string>

#include <openssl/sha.h>
#include <openssl/aes.h>
#include <openssl/evp.h>
#include <openssl/rand.h>

using namespace std;

string Cryptography::SHA256(const FString Message)
{
	const string Str = TCHAR_TO_UTF8(*Message);
	unsigned char Hash[SHA256_DIGEST_LENGTH];
	SHA256_CTX SHA256; SHA256_Init(&SHA256);
	SHA256_Update(&SHA256, Str.c_str(), Str.size());
	SHA256_Final(Hash, &SHA256);
	
	stringstream MessageHash;
	for(int i = 0; i < SHA256_DIGEST_LENGTH; i++) MessageHash << hex << setw(2) << setfill('0') << static_cast<int>(Hash[i]);

	return MessageHash.str();
}

void Cryptography::Encrypt(const TArray<uint8> InMessage, string& OutMessage, string& OutIv)
{
	TArray<uint8> Iv;
	Iv.AddUninitialized(16);
	RAND_bytes(Iv.GetData(), 16);
	
	EVP_CIPHER_CTX* Ctx = EVP_CIPHER_CTX_new();
	EVP_EncryptInit_ex(Ctx, EVP_aes_256_cbc(), nullptr, SharedSecretKey.GetData(), Iv.GetData());

	TArray<uint8> EncryptedMsg;
	EncryptedMsg.AddUninitialized(InMessage.Num() + EVP_CIPHER_CTX_block_size(Ctx));
	int32 Len;
	EVP_EncryptUpdate(Ctx, EncryptedMsg.GetData(), &Len,InMessage.GetData(), InMessage.Num());
	int32 CiphertextLen = Len;
	EVP_EncryptFinal_ex(Ctx, EncryptedMsg.GetData() + Len, &Len);
	CiphertextLen += Len;
	EncryptedMsg.SetNum(CiphertextLen);
	
	OutIv = TCHAR_TO_ANSI(*FBase64::Encode(Iv));
	OutMessage = TCHAR_TO_ANSI(*FBase64::Encode(EncryptedMsg));
	EVP_CIPHER_CTX_free(Ctx);
}

string Cryptography::Decrypt(const string InMessage, const string InIv)
{
	TArray<uint8> Message;
	FBase64::Decode(FString(InMessage.c_str()), Message);
	TArray<uint8> Iv;
	FBase64::Decode(FString(InIv.c_str()), Iv);
	
	EVP_CIPHER_CTX* Ctx = EVP_CIPHER_CTX_new();
	EVP_DecryptInit_ex(Ctx, EVP_aes_256_cbc(), nullptr, SharedSecretKey.GetData(), Iv.GetData());
	
	TArray<uint8> OutBytes;
	OutBytes.AddUninitialized(Message.Num() + EVP_CIPHER_CTX_block_size(Ctx));

	int32 Len;
	EVP_DecryptUpdate(Ctx, OutBytes.GetData(), &Len, Message.GetData(), Message.Num());
	EVP_DecryptFinal_ex(Ctx, OutBytes.GetData() + OutBytes.Num(), &Len);

	EVP_CIPHER_CTX_free(Ctx);
	return string(reinterpret_cast<char*>(OutBytes.GetData()), OutBytes.Num());
}

/* ========================== Diffie Hellman ========================== */

string Cryptography::GenerateKeyPair_DH()
{
	if(!DhKeyExchange)
	{
		DhKeyExchange = DH_new();
		constexpr unsigned char dh_p[] = {
			0xEB, 0x61, 0x99, 0x83, 0x8E, 0x4E, 0x30, 0xFD, 0x0F, 0xC0,
			0x4E, 0x1B, 0x69, 0xDB, 0x3D, 0x85, 0x59, 0x4D, 0xB8, 0x55,
			0x08, 0xAF, 0x9D, 0x5D, 0x83, 0xB7, 0x60, 0x20, 0xA0, 0xCE,
			0x22, 0x23, 0x8D, 0x00, 0x72, 0x06, 0xA9, 0xC0, 0x55, 0x01,
			0x31, 0xB9, 0xFB, 0x73, 0xD2, 0x1F, 0xF0, 0x0C, 0xFF, 0xA2,
			0xCE, 0x8F, 0xB1, 0xCA, 0x64, 0xA3, 0xBF, 0x46, 0x61, 0x45,
			0xB6, 0xD1, 0xD7, 0x2B};
		
		constexpr unsigned char dh_g[] = {0x02};
		DH_set0_pqg(DhKeyExchange, BN_bin2bn(dh_p, sizeof dh_p, nullptr), nullptr, BN_bin2bn(dh_g, sizeof dh_g, nullptr));
	}

	DH_generate_key(DhKeyExchange);
	return BN_bn2dec(DH_get0_pub_key(DhKeyExchange));
}

void Cryptography::ComputeSharedSecretKey_DH(const string PublicKey)
{
	BIGNUM* PubKey = BN_new();
	BN_dec2bn(&PubKey, PublicKey.c_str());
	
	SharedSecretKey.AddUninitialized(64);
	DH_compute_key(SharedSecretKey.GetData(), PubKey, DhKeyExchange);
	
	DH_free(DhKeyExchange);
	DhKeyExchange = nullptr;
	BN_free(PubKey);
	KeyExchangeHasBeenCompleted = true;
	GLog->Log("Successful secret key generation");
}

File: Server/src/main/java/com/server/Utils/Cryptography.java
package com.server.Utils;

import javax.crypto.*;
import javax.crypto.interfaces.DHPublicKey;
import javax.crypto.spec.DHParameterSpec;
import javax.crypto.spec.DHPublicKeySpec;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.math.BigInteger;
import java.security.*;
import java.security.spec.InvalidKeySpecException;
import java.util.Arrays;
import java.util.Base64;

public class Cryptography
{
    public Object[] Encrypt(String InMessage)
    {
        try
        {
            byte[] IV = new byte[16];
            new SecureRandom().nextBytes(IV);
            IvParameterSpec IvParameter = new IvParameterSpec(IV);

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, SharedSecretKey, IvParameter);
            byte[] EncryptedMsg = cipher.doFinal(InMessage.getBytes());
            return new Object[]{new String(Base64.getEncoder().encode(EncryptedMsg)), new String(Base64.getEncoder().encode(IV))};
        }
        catch (InvalidKeyException | NoSuchPaddingException | NoSuchAlgorithmException | IllegalBlockSizeException | BadPaddingException | InvalidAlgorithmParameterException e)
        {e.printStackTrace();}
        return null;
    }

    public byte[] Decrypt(String InMessage, String IV)
    {
        try
        {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, SharedSecretKey, new IvParameterSpec(Base64.getDecoder().decode(IV)));
            return cipher.doFinal(Base64.getDecoder().decode(InMessage));
        }
        catch (IllegalBlockSizeException | BadPaddingException | NoSuchPaddingException | NoSuchAlgorithmException | InvalidKeyException | InvalidAlgorithmParameterException e)
        {e.printStackTrace();}
        return null;
    }

    /* ========================== Diffie Hellman ========================== */
    public SecretKeySpec SharedSecretKey;
    private KeyAgreement DhKeyExchange;
    private static final byte[] dh_p = {
            (byte)0xEB, (byte)0x61, (byte)0x99, (byte)0x83, (byte)0x8E, (byte)0x4E, (byte)0x30, (byte)0xFD, (byte)0x0F, (byte)0xC0,
            (byte)0x4E, (byte)0x1B, (byte)0x69, (byte)0xDB, (byte)0x3D, (byte)0x85, (byte)0x59, (byte)0x4D, (byte)0xB8, (byte)0x55,
            (byte)0x08, (byte)0xAF, (byte)0x9D, (byte)0x5D, (byte)0x83, (byte)0xB7, (byte)0x60, (byte)0x20, (byte)0xA0, (byte)0xCE,
            (byte)0x22, (byte)0x23, (byte)0x8D, (byte)0x00, (byte)0x72, (byte)0x06, (byte)0xA9, (byte)0xC0, (byte)0x55, (byte)0x01,
            (byte)0x31, (byte)0xB9, (byte)0xFB, (byte)0x73, (byte)0xD2, (byte)0x1F, (byte)0xF0, (byte)0x0C, (byte)0xFF, (byte)0xA2,
            (byte)0xCE, (byte)0x8F, (byte)0xB1, (byte)0xCA, (byte)0x64, (byte)0xA3, (byte)0xBF, (byte)0x46, (byte)0x61, (byte)0x45,
            (byte)0xB6, (byte)0xD1, (byte)0xD7, (byte)0x2B};

    private static final BigInteger P = new BigInteger(1, dh_p);
    private static final BigInteger G = BigInteger.valueOf(2);
    public String GenerateKeyPair_DH()
    {
        try
        {
            KeyPairGenerator keypairGenerator = KeyPairGenerator.getInstance("DH");
            keypairGenerator.initialize(new DHParameterSpec(P, G));
            KeyPair keyPair = keypairGenerator.generateKeyPair();

            DHPublicKey dhPub = (DHPublicKey)keyPair.getPublic();

            DhKeyExchange = KeyAgreement.getInstance("DH");
            DhKeyExchange.init(keyPair.getPrivate());
            return String.valueOf(dhPub.getY());
        }
        catch (InvalidAlgorithmParameterException | NoSuchAlgorithmException | InvalidKeyException e)
        {e.printStackTrace();}
        return "";
    }

    public void ComputeSharedSecretKey_DH(String PublicKey)
    {
        try
        {
            DHPublicKeySpec dhPubKeySpecs = new DHPublicKeySpec(new BigInteger(PublicKey), P, G);
            KeyFactory kf = KeyFactory.getInstance("DH");
            DHPublicKey PubKey = (DHPublicKey) kf.generatePublic(dhPubKeySpecs);

            DhKeyExchange.doPhase(PubKey, true);
            SharedSecretKey = new SecretKeySpec(Arrays.copyOfRange(DhKeyExchange.generateSecret(), 0, 32), "AES");
            System.out.println("Successful secret key generation");
        }
        catch (NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException e)
        {e.printStackTrace();}
    }
}